### ISSUE TITLE
Handle common problems in `german-sysadmin`

### ISSUE THREAD

**Issue by angelikatyborska** [exercism/elixir-analyzer#169]
As @Cohen-Carlisle rightfully brought to our attention, the community solutions for this exercise are full of antipatterns, and people doing them aren't requesting mentoring, so they are not aware something is wrong. Some of those could be detected with the analyzer, but it might be very tricky to get them right. Let's discuss.

I am not going to provide links to specific solutions because this is a public discussion and I don't want to shame anyone in particular, I'll just copy them.

All ideas on how to detect those problems as broadly as possible without false positives are very welcome.

## Problem 1: `acc + [elem]`

Some examples:

With `Enum.reduce`

```elixir
    reducer = fn char, acc ->
      acc ++
        case char do
          ?ä -> 'ae'
          ?ö -> 'oe'
          ?ü -> 'ue'
          ?ß -> 'ss'
          ?_ -> '_'
          char when char in ?a..?z -> [char]
          _ -> ''
        end
    end
    Enum.reduce(username, '', reducer)
```

With tail-call recursion

```elixir
  def sanitize([head | tail], acc) do
    acc =
      case head do
        h when (h >= ?a and h <= ?z) or h == ?_ ->
          acc ++ [h]
        h when h == ?ä ->
          acc ++ 'ae'
        h when h == ?ö ->
          acc ++ 'oe'
        h when h == ?ü ->
          acc ++ 'ue'
        h when h == ?ß ->
          acc ++ 'ss'
        _ ->
          acc
      end
    sanitize(tail, acc)
  end
```

In general tail-call recursion is still a valid approach here if it doesn't have the `acc + [elem]` problem, but it might still be good enough to tell everyone to use simple recursion (less code).

## Problem 2: integer literals written as integers, without `?`

This one is very hard to tackle because both syntax variants produce exactly the same AST. Should we... run a regex on the string solution? 😱 

Example:
```elixir
  defp sanitize_char(char) do
    case char do
      252 -> 'ue'
      246 -> 'oe'
      228 -> 'ae'
      223 -> 'ss'
      _ -> char
    end
  end
```

## Problem 3: lots of repetition that could be avoided by returning the replacement list from `case`

This is not a biggie, but might be easy to detect?

Instead of:
```elixir
    case username do
      [] -> []
      [?ä | t] -> [?a, ?e | sanitize(t)]
      [?ü | t] -> [?u, ?e | sanitize(t)]
      [?ö | t] -> [?o, ?e | sanitize(t)]
      [?ß | t] -> [?s, ?s | sanitize(t)]
      [?_ | t] -> [?_ | sanitize(t)]
      [h | t] when h in ?a..?z -> [h | sanitize(t)]
      [_ | t] -> sanitize(t)
    end
```

It "should" be:
```elixir
    replacement = case username do
      [] -> []
      [?ä | t] -> [?a, ?e]
      [?ü | t] -> [?u, ?e]
      [?ö | t] -> [?o, ?e]
      [?ß | t] -> [?s, ?s]
      [?_ | t] -> [?_]
      [h | t] when h in ?a..?z -> [h]
      [_ | t] -> []
    end
    replacement ++ sanitize(t)
```

**Comment by jiegillet** [comment]
For problem 2, I say yes. We can introduce a new type of check like we discussed in #86. Something like

```elixir
source_check do
  find :any
  type :actionable
  comment "elixir.german-sysadmin.use_codepoints"

  check(source, _ast) do
    integers = ["252", "246", "228", "223"]
    Enum.any?(integers, &String.contains?(source, &1))
  end
end
``` 

I don't know hot to approach the other problems yet.

**Comment by angelikatyborska** [comment]
Thanks for the ping, I must have lost the notification email somewhere for the previous comment!

I see on https://github.com/exercism/elixir-analyzer/issues/86 that my opinion on:

> 4. Introducing a test check that can do ad-hoc operations on the raw code

was:

>  I don't put much value in the multiline string check, so I wouldn't do options 2-4 as it's a lot of work.

However, I do put a lot of value in using `?` in this exercise, so consider my opinion overall changed 🙂 

I like this:

```
source_check do
  find :any
  type :actionable
  comment "elixir.german-sysadmin.use_codepoints"

  check(source, _ast) do
    integers = ["252", "246", "228", "223"]
    Enum.any?(integers, &String.contains?(source, &1))
  end
end
```

It will open up the possibilities of custom exercise-specific checks, which is something that I have been missing a lot.
